numpy/ma/core.py

```diff
@@ -2837,6 +2837,13 @@ def __new__(cls, data=None, mask=nomask, dtype=None, copy=False,
             _data = ndarray.view(_data, type(data))
         else:
             _data = ndarray.view(_data, cls)
+
+        # Handle the case where data is not a subclass of ndarray, but
+        # still has the _mask attribute like MaskedArrays
+        if hasattr(data, '_mask') and not isinstance(data, ndarray):
+            _data._mask = data._mask
+            _data._sharedmask = True
```

Comment:
Was that actually there before, I think this is a bug-fix probably?  I feel there was some discussion about it being `_sharedmask = True` rather than `_data._sharedmask = True`?

Reply:
Here is the discussion thread: https://github.com/numpy/numpy/pull/19102#discussion_r639641117
with the previous code removed from that commit
```python
            # FIXME: should we set `_data._sharedmask = True`? 
            # Previously this set `_sharedmask = True`, but
            # that didn't actually do anything as the variable was unused!
```

I think this makes sense from the standpoint of when we are putting the incoming data._mask onto the new object, we should inform the original object that it is now sharing its mask with something else. I agree that it is not a full reversion in that sense, but I wasn't sure whether to call this a fix or revert PR, since the revert part is probably more important to get across. Happy to remove this and go back to a pure revert, or I can also add this "fix" piece as a separate commit if that would be more clear too.

Reply:
I am good with this (might rename the PR title, but that is all).  I just wanted to make sure that you are certain that it is correct (if we could add a test for it that would be awesome).
Basically, I am not deep into masked arrays, so just probing to build my trust in the fix ;).

Reply:
OK, good suggestion to dig into this a bit more @seberg :)
Adding `_data.sharedmask = True` would actually break the case where `keep_mask=False` and overrides the mask later, but doesn't update the `_sharedmask` attribute. I decided to leave this as purely a partial revert of that commit and added a new test for that case I found at the end here as well.

Another option is to fix this by adding in the `_data._sharedmask = True` here and forcing `sharedmask = False` in the if statement here: https://github.com/numpy/numpy/blob/f146ec16eed3f464f152ac9be4d51e33602f4e80/numpy/ma/core.py#L2847. But, I think that should be for a separate FIX PR if someone thinks it is necessary/desirable.

Reply:
> But, I think that should be for a separate FIX PR if someone thinks it is necessary/desirable.

Sounds good.  There are not many actively invested in masked arrays.  So if you think it is desirable than that is definitely "someone" :).